pkg/csvcopy/csvcopy_test.go

```diff
@@ -255,3 +345,171 @@ func TestWriteReportProgress(t *testing.T) {
+	require.NoError(t, err)
+	result, err := copier.Copy(context.Background(), reader)
+	assert.NoError(t, err)
+	assert.EqualValues(t, 4, result.RowsRead)
+
+	require.Contains(t, fs.Files, 1)
+	assert.Equal(t, fs.Files[1].String(), "24,qased,2.4\n24,qased,hello\n")
+	require.Contains(t, fs.Errors, 1)
+	assert.EqualValues(t, fs.Errors[1].(*ErrAtRow).RowAtLocation(), 3)
+	assert.EqualValues(t, fs.Errors[1].(*ErrAtRow).BatchLocation.ByteOffset, 26)
+}
+
+type MockErrorHandler struct {
+	Files  map[int]*bytes.Buffer
+	Errors map[int]error
+}
+
+func (fs *MockErrorHandler) HandleError(batch batch.Batch, reason error) error {
+	if fs.Files == nil {
+		fs.Files = map[int]*bytes.Buffer{}
+	}
+	if fs.Errors == nil {
+		fs.Errors = map[int]error{}
+	}
+	buf := &bytes.Buffer{}
+	_, err := buf.ReadFrom(&batch.Data)
+	if err != nil {
+		return err
+	}
+	fs.Files[int(batch.Location.StartRow)] = buf
+	fs.Errors[int(batch.Location.StartRow)] = reason
+	return nil
+}
+
+func TestFailedBatchHandlerFailure(t *testing.T) {
+	ctx := context.Background()
+
+	pgContainer, err := postgres.RunContainer(ctx,
+		testcontainers.WithImage("postgres:15.3-alpine"),
+		postgres.WithDatabase("test-db"),
+		postgres.WithUsername("postgres"),
+		postgres.WithPassword("postgres"),
+		testcontainers.WithWaitStrategy(
+			wait.ForLog("database system is ready to accept connections").
+				WithOccurrence(2).WithStartupTimeout(5*time.Second)),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Cleanup(func() {
+		if err := pgContainer.Terminate(ctx); err != nil {
+			t.Fatalf("failed to terminate pgContainer: %s", err)
+		}
+	})
+
+	connStr, err := pgContainer.ConnectionString(ctx, "sslmode=disable")
+	require.NoError(t, err)
+
+	conn, err := pgx.Connect(ctx, connStr)
+	require.NoError(t, err)
+	defer conn.Close(ctx)
+	_, err = conn.Exec(ctx, "create table public.metrics (device_id int, label text, value float8)")
+	require.NoError(t, err)
+
+	// Create a temporary CSV file
+	tmpfile, err := os.CreateTemp("", "example")
+	require.NoError(t, err)
+	defer os.Remove(tmpfile.Name())
+
+	// Write data to the CSV file
+	writer := csv.NewWriter(tmpfile)
+
+	data := [][]string{
+		// Batch 1
+		{"42", "xasev", "4.2"},
+		{"24", "qased", "2.4"},
+		// Batch 2
+		{"24", "qased", "2.4"},
+		{"24", "qased", "hello"},
+		// Batch 3
+		{"24", "qased", "2.4"},
+		{"24", "qased", "2.4"},
+	}
+
+	for _, record := range data {
+		if err := writer.Write(record); err != nil {
+			t.Fatalf("Error writing record to CSV: %v", err)
+		}
+	}
+
+	writer.Flush()
+
+	copier, err := NewCopier(connStr, "metrics", WithColumns("device_id,label,value"), WithBatchSize(2), WithBatchErrorHandler(func(batch batch.Batch, err error) error {
+		return fmt.Errorf("couldn't handle error %w", err)
+	}))
+	require.NoError(t, err)
+	reader, err := os.Open(tmpfile.Name())
+	require.NoError(t, err)
+	_, err = copier.Copy(context.Background(), reader)
+	assert.Error(t, err)
+	assert.ErrorContains(t, err, "couldn't handle error")
```

Comment:
why combine `require` and `assert` instead of using the same one for test? they don't do the same require aborts the test and assert continues

Reply:
you are right, this is provably me not paying much attention and copy pasting things. 

It is ok to mix them if done right, assert will continue and in case of a partial failure you will still get a full report. But here is clearly random :sweat_smile: